src/main/java/frc/robot/subsystems/swerve/states/heading/HeadingStabilizer.java

```diff
@@ -0,0 +1,43 @@
+package frc.robot.subsystems.swerve.states.heading;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.constants.MathConstants;
+import frc.robot.subsystems.swerve.SwerveConstants;
+
+public class HeadingStabilizer {
+
+	private final PIDController headingController;
+	private boolean targetLocked;
+
+	public HeadingStabilizer(SwerveConstants constants) {
+		this.headingController = new PIDController(
+			constants.rotationDegreesPIDController().getP(),
+			constants.rotationDegreesPIDController().getI(),
+			constants.rotationDegreesPIDController().getD()
+		);
+		headingController.enableContinuousInput(MathConstants.HALF_CIRCLE.unaryMinus().getDegrees(), MathConstants.HALF_CIRCLE.getDegrees());
+
+		this.targetLocked = false;
+	}
+
+	public void lockTarget() {
+		targetLocked = true;
+	}
+
+	public void unlockTarget() {
+		targetLocked = false;
+	}
+
+	public void setTargetHeading(Rotation2d targetHeading) {
+		if (!targetLocked) {
+			headingController.reset();
+			headingController.setSetpoint(targetHeading.getDegrees());
+		}
+	}
+
+	public Rotation2d calculatePIDOutput(Rotation2d currentHeading) {
+		return Rotation2d.fromDegrees(headingController.calculate(currentHeading.getDegrees()));
+	}
```

Comment:
since when pid output is Rotation2d? i thought it returns the power needed to apply to get to a certain target

Reply:
it returns velocity.
PID in general returns what you want depends on the case